examples/mamba.py

```diff
@@ -0,0 +1,472 @@
+        + math.log(dt_min)
+      )
+      .exp()
+      .maximum(dt_init_floor)
+    )
+    inv_dt = (
+      dt + (-((-dt).exp() - Tensor.ones(*dt.shape))).log()
+    )  # TODO: implement torch.expm1?
+
+    self.dt_proj.bias.assign(inv_dt)
+
+    # S4D real initialization
+    self.A_log = (
+      Tensor.arange(1, self.d_state + 1).repeat([self.d_inner, 1]).contiguous().log()
+    )
+
+    # D "skip" parameter
+    self.D = Tensor.ones(self.d_inner)  # Keep in fp32
+
+    self.out_proj = nn.Linear(self.d_inner, self.dim, bias=bias)
+
+  def __call__(self, hidden_states: Tensor, inference_params=None):
+    batch, seqlen, dim = hidden_states.shape
+
+    conv_state, ssm_state = None, None
+    if inference_params is not None:
+      conv_state, ssm_state = self._get_states_from_cache(inference_params, batch)
+      if inference_params.seqlen_offset > 0:
+        # The states are updated inplace
+        out, _, _ = self.step(hidden_states[:, -1:, :], conv_state, ssm_state)
+        return out
+
+    xz = rearrange(
+      self.in_proj.weight @ rearrange(hidden_states, "b l d -> d (b l)"),
+      "d (b l) -> b d l",
+      l=seqlen,
+    )
+
+    if self.in_proj.bias is not None:
+      xz = xz + rearrange(self.in_proj.bias, "d -> d 1")
+
+    A = -self.A_log.exp()
+    x, z = xz.chunk(2, dim=1)
+    # Compute short convolution
+    if conv_state is not None:
+      conv_state.assign(x[:, :, -self.d_conv :])  # Update state (B D W)
+      x = self.conv1d(x)[..., :seqlen].swish()
+
+    x_dbl = self.x_proj(rearrange(x, "b d l -> (b l) d"))
+    dt = x_dbl[:, : self.dt_rank]
+    B = x_dbl[:, self.dt_rank : (self.dt_rank + self.d_state)]
+    C = x_dbl[:, (self.dt_rank + self.d_state) :]
+    
+    #print(f"dt before: {dt.shape}")
+    dt = self.dt_proj.weight @ dt.T
+    #print(f"dt mul: {dt.shape}")
+    dt = rearrange(dt, "d (b l) -> b d l", l=seqlen)
+    #print(f"dt rearranged: {dt.shape}")
+
+    B = rearrange(B, "(b l) dstate -> b dstate l", l=seqlen).contiguous()
+    C = rearrange(C, "(b l) dstate -> b dstate l", l=seqlen).contiguous()
+
+    y = selective_scan_ref(  # TODO: actually implement selective_scan_fn
+      x,
+      dt,
+      A,
+      B,
+      C,
+      self.D,
+      z=z,
+      delta_bias=self.dt_proj.bias,
+      delta_softplus=True,
+      return_last_state=ssm_state is not None,
+    )
+
+    if ssm_state is not None:
+      y, last_state = y
+      ssm_state.assign(last_state)
+
+    y = rearrange(y, "b d l -> b l d")
+    out = self.out_proj(y)
+
+    return out
+
+  def step(self, hidden_states, conv_state, ssm_state):
+    #print(f"hidden_states is {hidden_states.shape}")
+    assert (
+      hidden_states.shape[1] == 1
+    ), f"Only support decoding with 1 token at a time for now, attempted {hidden_states.shape[1]}"
+    xz = self.in_proj(hidden_states.squeeze(1))  # (B 2D)
+    x, z = xz.chunk(2, dim=-1)  # (B D)
+
+    # Conv step
+    conv_state = conv_state[:, :, 1:].cat(x.unsqueeze(-1), dim=-1)
```

Comment:
The problem is on this line
It creates a new object instead of updating in-place
But the cache keeps reference to the old object
So every time this function is called, this array is empty (all zeros)

Reply:
It can optionally just update the cache with the new object

Reply:
This solves the problem:
```python
conv_state.assign(conv_state[:, :, 1:].cat(x.unsqueeze(-1), dim=-1))
```